### PR TITLE
Migrate CPP SDK from FieldValue to Protobuf

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This changes the internals of the Firestore C++ to use the updated code on iOS, which drops `model::FieldValue` in favor of using Protobuf directly.

I did not port the UserDataReader/UserDataWriter refactor, but we can do that as a follow up.

This will not build until the next iOS Release.

### CODE REVIEW

File: firestore/src/main/user_data_converter_main.cc [hunk 1]
```diff
@@ -273,43 +291,63 @@ model::FieldValue::Array UserDataConverter::ParseArray(
...
-    absl::optional<model::FieldValue> parsed =
-        ParseData(value, context.ChildContext(key));
-    if (parsed) {
-      result = result.insert(key, std::move(parsed).value());
+    pb_size_t index = 0;
+    for (const auto& kv : input) {
+      const std::string& key = kv.first;
+      const FieldValue& value = kv.second;
+
+      auto parsedValue = ParseData(value, context.ChildContext(key));
```
Comment (var-const) [inline firestore/src/main/user_data_converter_main.cc:341]: Nit: `s/parsedValue/parsed_value/`.

File: firestore/src/main/user_data_converter_main.cc [hunk 2]
```diff
@@ -273,43 +291,63 @@ model::FieldValue::Array UserDataConverter::ParseArray(
...
+      if (parsedValue) {
+        result->map_value.fields[index].key = nanopb::MakeBytesArray(key);
+        result->map_value.fields[index].value = *parsedValue->release();
+        ++index;
+      }
     }
   }
 
-  return model::ObjectValue::FromMap(std::move(result));
+  return std::move(result);
```
Comment (var-const) [inline firestore/src/main/user_data_converter_main.cc:350]: It's not necessary (and sometimes can be slightly harmful to performance) to move when returning local variables.

File: firestore/src/main/document_snapshot_main.cc [hunk 3]
```diff
@@ -84,95 +92,111 @@ FieldValue DocumentSnapshotInternal::GetValue(
...
-    const model::FieldValue::Map& object, ServerTimestampBehavior stb) const {
+    const google_firestore_v1_MapValue& object,
+    ServerTimestampBehavior stb) const {
   MapFieldValue result;
-  for (const auto& kv : object) {
-    result[kv.first] = ConvertAnyValue(kv.second, stb);
+  for (pb_size_t i = 0; i < object.fields_count; ++i) {
+    std::string key = MakeString(object.fields[i].key);
+    const google_firestore_v1_Value& value = object.fields[i].value;
+    result[key] = ConvertAnyValue(value, stb);
```
Comment (var-const) [inline firestore/src/main/document_snapshot_main.cc:113]: Nit: you can move `key`.

File: firestore/src/main/document_snapshot_main.cc [hunk 4]
```diff
@@ -84,95 +92,111 @@ FieldValue DocumentSnapshotInternal::GetValue(
...
+    const google_firestore_v1_Value& reference) const {
+  std::string ref = MakeString(reference.reference_value);
+  DatabaseId database_id = DatabaseId::FromName(ref);
+  DocumentKey key = DocumentKey::FromName(ref);
+
+  SIMPLE_HARD_ASSERT(database_id == firestore_internal()->database_id(),
+                     "Converted reference is from another database");
 
-  api::DocumentReference api_reference{reference.key(), snapshot_.firestore()};
+  api::DocumentReference api_reference{key, snapshot_.firestore()};
```
Comment (var-const) [inline firestore/src/main/document_snapshot_main.cc:177]: You can probably move `key`.

File: firestore/src/main/query_main.cc [hunk 5]
```diff
@@ -190,10 +210,10 @@ core::Bound QueryInternal::ToBound(
       SimpleThrowInvalidArgument(message);
     }
 
-    components.push_back(std::move(value).value());
+    components->values[i] = *DeepClone(*value).release();
```
Comment (var-const) [inline firestore/src/main/query_main.cc:213]: Question: why is `DeepClone` necessary here?

File: firestore/src/main/query_main.cc [hunk 6]
```diff
@@ -210,29 +230,34 @@ core::Bound QueryInternal::ToBound(
...
-  if (from.type() != model::FieldValue::Type::String) {
+Message<google_firestore_v1_Value> QueryInternal::ConvertDocumentId(
+    const Message<google_firestore_v1_Value>& from,
+    const core::Query& internal_query) const {
+  if (GetTypeOrder(*from) != TypeOrder::kString) {
     SimpleThrowInvalidArgument(
         "Invalid query. Expected a string for the document ID.");
   }
-  const std::string& document_id = from.string_value();
+  const std::string& document_id = MakeString(from->string_value);
```
Comment (var-const) [inline firestore/src/main/query_main.cc:260]: Please make `document_id` a non-reference variable to avoid attaching a reference to a temporary.

**Review by var-const — CHANGES_REQUESTED** [review]

File: firestore/src/main/document_snapshot_main.cc (hunk 3 above)
Comment (schmidt-sebastian) [reply firestore/src/main/document_snapshot_main.cc]: Done

File: firestore/src/main/field_value_main.cc [hunk 7]
```diff
@@ -16,39 +17,68 @@ namespace firestore {
...
 using Type = FieldValue::Type;
 
 }  // namespace
 
 // Constructors
 
-FieldValueInternal::FieldValueInternal(bool value)
-    : type_{Type::kBoolean}, value_{model::FieldValue::FromBoolean(value)} {}
+FieldValueInternal::FieldValueInternal(bool value) : type_{Type::kBoolean} {
+  auto proto = absl::get<SharedMessage<google_firestore_v1_Value>>(value_);
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firestore/src/main/field_value_main.cc]: I added the helper method, which reduces the number of `SharedMessage<google_firestore_v1_Value>` types significantly. I do not think we need an alias anymore. 

Also changes all access to references.

File: firestore/src/main/field_value_main.cc [hunk 8]
```diff
@@ -60,37 +90,48 @@ FieldValueInternal::FieldValueInternal(MapFieldValue value)
...
   SIMPLE_HARD_ASSERT(type_ == Type::kString);
-  return absl::get<model::FieldValue>(value_).string_value();
+  return MakeString(absl::get<SharedMessage<google_firestore_v1_Value>>(value_)
+                        ->string_value);
 }
 
 const uint8_t* FieldValueInternal::blob_value() const {
   SIMPLE_HARD_ASSERT(type_ == Type::kBlob);
-  return absl::get<model::FieldValue>(value_).blob_value().data();
+  auto value =
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firestore/src/main/field_value_main.cc]: Done

File: firestore/src/main/field_value_main.cc [hunk 9]
```diff
@@ -60,37 +90,48 @@ FieldValueInternal::FieldValueInternal(MapFieldValue value)
...
+      absl::get<SharedMessage<google_firestore_v1_Value>>(value_)->bytes_value;
+  return value ? value->bytes : nullptr;
 }
 
 size_t FieldValueInternal::blob_size() const {
   SIMPLE_HARD_ASSERT(type_ == Type::kBlob);
-  return absl::get<model::FieldValue>(value_).blob_value().size();
+  auto value =
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firestore/src/main/field_value_main.cc]: Done

File: firestore/src/main/query_main.cc (hunk 5 above)
Comment (schmidt-sebastian) [reply firestore/src/main/query_main.cc:213]: The memory is owned by the DocumentSnapshot and we need to move ownership to the query bound. See here: https://github.com/firebase/firebase-ios-sdk/blob/e8f419baf275899068e915a6c2ecb4ae1e43d264/Firestore/Source/API/FIRQuery.mm#L584

File: firestore/src/main/query_main.cc (hunk 6 above)
Comment (schmidt-sebastian) [reply firestore/src/main/query_main.cc]: Done.

File: firestore/src/main/user_data_converter_main.cc (hunk 1 above)
Comment (schmidt-sebastian) [reply firestore/src/main/user_data_converter_main.cc]: Fixed

File: firestore/src/main/user_data_converter_main.cc (hunk 2 above)
Comment (schmidt-sebastian) [reply firestore/src/main/user_data_converter_main.cc]: Fixed

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the quick review. I addressed your feedback, but I did not verify the build. I will wait until this is ready for merging until I assign this back.

**Review by var-const — DISMISSED** [review]

File: firestore/integration_test_internal/CMakeLists.txt [hunk 10]
```diff
@@ -204,6 +204,7 @@ include_directories(src)
 include_directories(${FIREBASE_CPP_SDK_DIR})
 # Additional public headers from Firestore core SDK.
 include_directories(${PROJECT_BINARY_DIR}/bin/external/src/firestore)
+include_directories(${PROJECT_BINARY_DIR}/bin/external/src/nanopb)
```
Comment (a-maurice) [inline firestore/integration_test_internal/CMakeLists.txt:207]: I think you also need to add the following line:
include_directories(${PROJECT_BINARY_DIR}/bin/external/src/firestore/Firestore/Protos/nanopb)

**Review by var-const — DISMISSED** [review]

File: Android/firebase_dependencies.gradle [hunk 11]
```diff
@@ -37,7 +37,7 @@ def firebaseDependenciesMap = [
                  'com.google.flatbuffers:flatbuffers-java:1.12.0',
                  'com.google.android.gms:play-services-base:17.6.0'],
   'performance' : ['com.google.firebase:firebase-perf:20.0.2'],
-  'remote_config' : ['com.google.firebase:firebase-config:21.0.0',
+  'remote_config' : ['com.google.firebase:firebase-config:21.0.1',
```
Comment (var-const) [inline Android/firebase_dependencies.gradle:40]: Is this change intentional?

File: scripts/gha/integration_testing/test_validation.py [hunk 12]
```diff
@@ -80,7 +80,7 @@ def validate_results_unity(log_text):
     # After the result string comes a list of failing testapps.
     summary = log_text[results_match.start():]
   else:
-    summary = _tail(log_text, 15)
+    summary = _tail(log_text, 1500000)
```
Comment (var-const) [inline scripts/gha/integration_testing/test_validation.py:83]: Should these be reverted?

**Review by var-const — CHANGES_REQUESTED** [review]

File: .github/workflows/integration_tests.yml [hunk 13]
```diff
@@ -14,7 +14,7 @@ on:
         required: true
       apis:
         description: 'CSV of apis to build and test'
-        default: 'admob,analytics,auth,database,dynamic_links,firestore,functions,installations,messaging,remote_config,storage'
+        default: 'firestore'
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply .github/workflows/integration_tests.yml]: Reverted

File: scripts/gha/integration_testing/test_validation.py (hunk 12 above)
Comment (schmidt-sebastian) [reply scripts/gha/integration_testing/test_validation.py]: Reverted

File: Android/firebase_dependencies.gradle (hunk 11 above)
Comment (schmidt-sebastian) [reply Android/firebase_dependencies.gradle:40]: Yes, this is part of the version update that we want to release.

**Review by a-maurice — APPROVED** [review]

**Review by var-const — APPROVED** [review]
